File: examples/python/apigateway_deploy.py
"""Despliega una API REST a un stage, dejándola accesible por HTTP.

Uso: python3 apigateway_deploy.py <api-id> [stage]
"""
import sys

import boto3

apigateway = boto3.client(
    "apigateway",
    endpoint_url="http://localhost:4566",
    aws_access_key_id="test",
    aws_secret_access_key="test",
    region_name="us-east-1",
)


def main():
    if len(sys.argv) < 2:
        raise SystemExit("Uso: python3 apigateway_deploy.py <api-id> [stage]")
    api_id = sys.argv[1]
    stage = sys.argv[2] if len(sys.argv) > 2 else "dev"

    # CreateDeployment congela una "foto" de los recursos/métodos actuales de la API y la
    # publica bajo un stage. Sin este paso, los métodos configurados con apigateway_put_method
    # solo existen en la definición de la API, pero no son alcanzables por HTTP todavía.
    apigateway.create_deployment(restApiId=api_id, stageName=stage)

    invoke_url = f"http://localhost:4566/restapis/{api_id}/{stage}/_user_request_"
    print(f'API desplegada en el stage "{stage}".')
    print("URL de invocación:", invoke_url)
    print(f"Prueba con: curl {invoke_url}/tareas")


if __name__ == "__main__":
    main()

File: examples/java/ApiGatewayDeploy.java
// Despliega una API REST a un stage, dejándola accesible por HTTP.
// Uso: java ApiGatewayDeploy <api-id> [stage]
import java.net.URI;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.apigateway.ApiGatewayClient;

public class ApiGatewayDeploy {
  public static void main(String[] args) {
    if (args.length < 1) {
      throw new IllegalArgumentException("Uso: java ApiGatewayDeploy <api-id> [stage]");
    }
    String apiId = args[0];
    String stage = args.length > 1 ? args[1] : "dev";

    ApiGatewayClient apigateway = ApiGatewayClient.builder()
        .endpointOverride(URI.create("http://localhost:4566"))
        .region(Region.US_EAST_1)
        .credentialsProvider(
            StaticCredentialsProvider.create(AwsBasicCredentials.create("test", "test")))
        .build();

    try {
      // CreateDeployment congela una "foto" de los recursos/métodos actuales de la API y la
      // publica bajo un stage. Sin este paso, los métodos configurados con ApiGatewayPutMethod
      // solo existen en la definición de la API, pero no son alcanzables por HTTP todavía.
      apigateway.createDeployment(req -> req.restApiId(apiId).stageName(stage));

      String invokeUrl = "http://localhost:4566/restapis/" + apiId + "/" + stage + "/_user_request_";
      System.out.println("API desplegada en el stage \"" + stage + "\".");
      System.out.println("URL de invocación: " + invokeUrl);
      System.out.println("Prueba con: curl " + invokeUrl + "/tareas");
    } finally {
      apigateway.close();
    }
  }
}
